backend/mttnet/site.go

```diff
@@ -0,0 +1,888 @@
+	if proxied {
+		retValue, ok := res.(*site.GetPathResponse)
+		if !ok {
+			return &site.GetPathResponse{}, fmt.Errorf("Format of proxied return value not recognized")
+		}
+		return retValue, nil
+	}
+	if in.Path == "" {
+		return &site.GetPathResponse{}, fmt.Errorf("Invalid path")
+	}
+	n, ok := srv.Node.Get()
+	if !ok {
+		return &site.GetPathResponse{}, fmt.Errorf("Node not ready yet")
+	}
+	conn, cancel, err := n.vcs.DB().Conn(ctx)
+	if err != nil {
+		return &site.GetPathResponse{}, fmt.Errorf("Cannot connect to internal db: %w", err)
+	}
+	defer cancel()
+	record, err := sitesql.GetWebPublicationRecordByPath(conn, in.Path)
+	if err != nil {
+		return &site.GetPathResponse{}, fmt.Errorf("Could not get record for path [%s]: %w", in.Path, err)
+	}
+	ret, err := srv.localFunctions.GetPublication(ctx, &site.GetPublicationRequest{
+		DocumentId: record.Document.ID.String(),
+		LocalOnly:  true,
+	})
+	if err != nil {
+		return &site.GetPathResponse{}, fmt.Errorf("Could not get local document although was found in the list of published documents: %w", err)
+	}
+	return &site.GetPathResponse{Publication: ret}, err
+}
+
+func getRemoteSiteFromHeader(ctx context.Context) (string, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", fmt.Errorf("There is no metadata provided in context")
+	}
+	token := md.Get(string(MttHeader))
+	if len(token) != 1 {
+		return "", fmt.Errorf("Header [%s] not found in metadata", MttHeader)
+	}
+	return token[0], nil
+}
+
+// getRemoteID gets the remote peer id if there is an opened p2p connection between them with context ctx.
+func getRemoteID(ctx context.Context) (cid.Cid, error) {
+	info, ok := rpcpeer.FromContext(ctx)
+	if !ok {
+		return cid.Cid{}, fmt.Errorf("BUG: no peer info in context for grpc")
+	}
+
+	pid, err := peer.Decode(info.Addr.String())
+	if err != nil {
+		return cid.Cid{}, err
+	}
+
+	return peer.ToCid(pid), nil
+}
+
+func randStr(length int) string {
+	randomBytes := make([]byte, 32)
+	_, err := rand.Read(randomBytes)
+	if err != nil {
+		panic(err)
+	}
+	return base32.StdEncoding.EncodeToString(randomBytes)[:length]
+}
+
+// Client dials a remote peer if necessary and returns the RPC client handle.
+func (srv *Server) Client(ctx context.Context, device cid.Cid) (site.WebSiteClient, error) {
+	n, ok := srv.Node.Get()
+	if !ok {
+		return nil, fmt.Errorf("Node not ready yet")
+	}
+	pid, err := peer.FromCid(device)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := n.Connect(ctx, n.p2p.Peerstore().PeerInfo(pid)); err != nil {
+		return nil, err
+	}
+	return n.client.DialSite(ctx, pid)
+}
+
+func (srv *Server) checkPermissions(ctx context.Context, requiredRole site.Member_Role, params ...interface{}) (cid.Cid, bool, interface{}, error) {
+	n, ok := srv.Node.Get()
+	if !ok {
+		return cid.Cid{}, false, nil, fmt.Errorf("Node not ready yet")
+	}
+
+	remoteHostname, err := getRemoteSiteFromHeader(ctx)
+	if err != nil && srv.hostname == "" { // no headers and not a site
```

Comment:
@juligasa aren't these separate conditions? If server is not a site is one thing, and if `getRemoteSiteFromHeader` returns an error is another thing?

Reply:
well, since `getRemoteSiteFromHeader` return an error if it does not find the required headers (or if it does not find any metadata at all) Then only in the case that the client did not provide headers and the server its not a site we return an error. Note that in the case the client does not provide a header, but we are a site, is a local call from the nextjs to the site, which should be allowed.